mono/metadata/native-library.c

```diff
@@ -408,126 +392,393 @@ mono_lookup_pinvoke_call_internal (MonoMethod *method, MonoError *error)
+
+	if (mono_runtime_get_no_exec ())
+		return NULL;
+
+	HANDLE_FUNCTION_ENTER ();
+
+	MonoString *scope_obj = mono_string_new_checked (domain, scope, error);
+	goto_if_nok (error, leave);
+	MonoReflectionAssemblyHandle assembly_handle = mono_assembly_get_object_handle (domain, assembly, error);
+	goto_if_nok (error, leave);
+
+	gpointer gchandle = GUINT_TO_POINTER (alc->gchandle);
+	gpointer args [4];
+	args [0] = scope_obj;
+	args [1] = MONO_HANDLE_RAW (assembly_handle);
+	args [2] = &gchandle;
+	args [3] = &lib;
+	mono_runtime_invoke_checked (resolve, NULL, args, error);
+	goto_if_nok (error, leave);
+
+	native_library_lock ();
+	result = netcore_handle_lookup (lib);
+	native_library_unlock ();
+
+leave:
+	HANDLE_FUNCTION_RETURN_VAL (result);
+}
+
+static MonoDl *
+netcore_resolve_with_resolving_event_nofail (MonoAssemblyLoadContext *alc, MonoAssembly *assembly, const char *scope)
+{
+	MonoDl *result = NULL;
+	ERROR_DECL (error);
+
+	result = netcore_resolve_with_resolving_event (alc, assembly, scope, error);
+	if (!is_ok (error))
+		mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_ASSEMBLY, "Error while invoking ALC ResolvingUnmangedDll(\"%s\") event: '%s'", scope, mono_error_get_message (error));
+
+	mono_error_cleanup (error);
+
+	return result;
+}
+
+static MonoDl *
+netcore_lookup_native_library (MonoAssemblyLoadContext *alc, MonoImage *image, const char *scope, guint32 flags)
+{
+	MonoDl *module = NULL;
+	MonoAssembly *assembly = mono_image_get_assembly (image);
+	char *error_msg = NULL;
+
+	MONO_REQ_GC_UNSAFE_MODE;
+
+	mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_DLLIMPORT, "DllImport attempting to load: '%s'.", scope);
+
+	// We allow a special name to dlopen from the running process namespace, which is not present in CoreCLR
+	if (strcmp (scope, "__Internal") == 0) {
+		if (!internal_module)
+			internal_module = mono_dl_open (NULL, MONO_DL_LAZY, &error_msg);
+		module = internal_module;
+
+		if (!module) {
+			mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_DLLIMPORT, "DllImport error loading library '__Internal': '%s'.", error_msg);
+			g_free (error_msg);
 		}
-		mono_image_unlock (image);
+
+		return module;
 	}
 
-	mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_DLLIMPORT,
-				"DllImport searching in: '%s' ('%s').", new_scope, found_name);
+	/*
+	 * Try these until one of them succeeds:
+	 *
+	 * 1. Check the cache in the active ALC.
+	 *
+	 * 2. Call the DllImportResolver on the active assembly.
+	 *
+	 * 3. Call LoadUnmanagedDll on the active ALC.
+	 *
+	 * 4. Check the global cache.
+	 *
+	 * 5. Run the unmanaged probing logic.
+	 *
+	 * 6. Raise the ResolvingUnmanagedDll event on the active ALC.
+	 *
+	 * 7. Return NULL.
+	 */
 
-	addr = pinvoke_probe_for_symbol (module, piinfo, new_import, &error_msg);
+	mono_alc_pinvoke_lock (alc);
+	module = (MonoDl *)g_hash_table_lookup (alc->pinvoke_scopes, scope);
+	native_library_lock ();
+	if (netcore_check_blacklist (module)) {
+		g_hash_table_remove (alc->pinvoke_scopes, scope);
+		module = NULL;
+	}
```

Comment:
oh god.  that's... a behavior.  So if I call `NativeLibrary.FreeLib` we have to make sure that any P/Invokes we haven't called before have to break in some reasonable way.

What if I already called the P/Invoke before and now I pull the NativeLibrary rug from underneath myself?  Do I just crash next time I call it?

I hope there's a test for this.

Reply:
Unloading shared libraries is a not very frequently used functionality, I think in C calling the functions after unloading will just crash or do something random. Rewriting the pinvokes to do something else would be nice but not sure its worth the effort.

Reply:
We need some sort of alc cache, and with nativelibrary.free we have to be able to invalidate it. I went between a couple designs, and ended up with the blacklist, but it's definitely messy.

There are almost no tests for any of this, and I don't feel comfortable merging it until I've written some, but at least there's an implementation to work off. In hindsight, I probably should have chunked this up better, and merging some of the earlier prs first while I iterate on this is an option. 

I have more work I'll be pushing tomorrow and so hopefully it'll be a bit clearer. This is going to stay a draft until we chat on Monday.